File: cpp/coinChange.cpp
#include <vector>
#include <limits.h>
using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount + 1, INT_MAX - 1);
        dp[0] = 0;
        
        for (int coin : coins) {
            for (int i = coin; i <=amount; i++) {
                dp[i] = min(dp[i], dp[i - coin] + 1);
            }
        }
        return (dp[amount] != INT_MAX - 1)? dp[amount] : -1;
    }
};

File: rs/coinChange.rs
impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let mut dp:Vec<i32> = vec![i32::MAX - 1; (amount + 1) as usize];
        dp[0] = 0;

        for coin in coins {
            for i in coin..=amount {
                dp[i as usize] = dp[i as usize].min(dp[(i - coin) as usize] + 1);
            }
        }

        if dp[amount as usize] != i32::MAX - 1 {return dp[amount as usize]}
        -1
    }
}